docs/ARCHITECTURE.md

```diff
@@ -0,0 +1,64 @@
+
+- [Etcher Image Stream](https://github.com/resin-io-modules/etcher-image-stream)
+
+The goal of this project is to convert any kind if input into a readable stream representing the image so it can be plugged to [etcher-image-write](https://github.com/resin-io-modules/etcher-image-write). Inputs that this module might handle could be, for example: a simple image file, a URL to an image, a compressed image, an image inside a ZIP archive, etc. Together with [etcher-image-write](https://github.com/resin-io-modules/etcher-image-write), these modules are the building blocks needed to take an image representation to the user's device, the "Etcher's backend".
+
+- [Drivelist](https://github.com/resin-io-modules/drivelist)
+
+As the name implies, this module's duty is to detect the connected drives uniformly in all major operating systems, along with valuable metadata, like if a drive is removable or not, to prevent users from trying to write an image to a system drive.
+
+- [Etcher](https://github.com/resin-io/etcher)
+
+This is the *"main repository"*, from which you're reading this from, which is basically the front-end and glue for all previously listed projects.
+
+Front-ends
+----------
+
+The main repository consists of the implementation of the Etcher CLI and the Etcher GUI (the desktop application), located at [`lib/cli/`](https://github.com/resin-io/etcher/tree/master/lib/cli) and [`lib/gui/`](https://github.com/resin-io/etcher/tree/master/lib/gui), respectively.
+
+In fact, the only front-end that interacts directly with Etcher's backend is the CLI. The GUI merely forks the CLI and communicates with its child process to get state information. 
+
+In this sense, you can consider the GUI as being the front-end to the CLI, which is in turn the front-end to the actual image writing functionality.
+
+As a way to simplify how the GUI forks the CLI in a packaged and distributed context, both the CLI and GUI share the same application entry point. This means that the same Etcher binary can behave as CLI or GUI as needed.
+
+## Process communication
+
+As mentioned before, the Etcher GUI forks the CLI and retrieves information from it to update its state. In order to accomplish this, the Etcher CLI contains certain features to ease communication:
+
+- [Well-documented exit codes.](https://github.com/resin-io/etcher/blob/master/lib/src/exit-codes.js)
+- A `--robot` option, which causes the Etcher CLI to output state in a way that can be easily machine-parsed.
+- An `--ipc` option, which causes the Etcher CLI to use [NodeJS's built-in IPC channel](https://nodejs.org/api/child_process.html#child_process_child_send_message_sendhandle_options_callback) to emit messages from the child to the parent.
+
+GUI fifty-thousand foot view
+----------------------------
+
+Given the event oriented nature of desktop application, it can be hard to follow what's going on without getting deep in the details.
```

Comment:
nature of desktop application -> nature of desktop applications or nature of a desktop application.